Add ability to use variables from grid file in input file expressions

ZedThree opened the pull request:
This adds the ability to read fields (2D and 3D) and doubles\* from the grid file (`mesh:file`) and use them in expressions in the input file, for example to use coordinates directly from the grid generator:

```ini
[input:grid_variables]
rho = field2d
theta = field2d
scale = boutreal

[mesh]
B = (scale / rho) * cos(theta)
```

_\* ints can also be read, but the expression parser only handles doubles_

---

This could also be used for something like region labels to the input file by having hypnotoad make a field which is 1.0 in that region and 0.0 elsewhere:

```python
region_pfr = 1.0 if is_in_pfr else 0.0
```

and using like

```ini
[input:grid_variables]
region_pfr = field2d

[some_source]
function = region_pfr * cos(x) * whatever
```

We could probably also do this in BOUT++ directly, making fields for each boundary region.

ZedThree committed:
Remove unused `Context` constructor

ZedThree committed:
Add cell indices to `generator::Context`

ZedThree committed:
Add ability to use variables from grid files in input expressions

ZedThree committed:
tests: Only remove temp files if test passes

ZedThree committed:
Enable reading `Field2D`s as well as `Field3D`s for expressions

ZedThree committed:
Refactor reading grid variables for input expressions

ZedThree committed:
Apply clang-format changes

dschwoerer reviewed (dismissed):
Looks good, thanks for cleaning up 👍

bendudson reviewed (dismissed):
Thanks @ZedThree ! This looks really useful. I think this is good as-is, but for future if this becomes widely used:
1. These expressions are in the global scope rather than being in the `mesh` namespace.
2. A nice-to-have might be renaming imports analogous to "import x as y" to avoid name conflicts.
These may over-complicate things, so I think this simplest implementation is good.

bendudson committed:
FieldFactory: Use fieldmesh not localmesh

The constructor argument `localmesh` can be null but in that
case `fieldmesh` is set to `bout::globals::mesh`.

bendudson committed:
GridVariable: Load grid variables lazily

`BoutMesh::load` creates a FieldFactory to read variables such as
mesh sizes. This then tries to load grid variables, which in turn
starts trying to create coordinates, use regions etc., none of which
has been constructed at that point.

This defers reading until it's needed, at which point it reads the
variable once and re-uses in subsequent calls.

bendudson commented:
Hey @ZedThree ! This functionality is really useful :) I had to make the loading lazy to avoid issues where a `FieldFactory` is created during `BoutMesh::load` and before `mesh->get` works.

bendudson committed:
FieldData: Don't getCoordinates in constructor

No need to get coordinates before they are needed. Calling
this can result in problems if mesh is not fully constructed.

bendudson committed:
GridVariable: Check mesh->get return

Not strictly necessary because variables are checked
before creating a GridVariable, but doesn't hurt.

bendudson commented:
Sorry, I force-pushed to remove a bad merge with `next`. It introduced all kinds of exciting memory errors.

bendudson committed:
FieldGenerators: Fixes and tidying

Missing argument to exception, misc. headers.

bendudson committed:
GridVariable unit tests: Fix unit test

Unit tests now include Y guard cells.

bendudson committed:
Field2D/3D: Defer sizing until allocation

FieldFactory can create Fields when bout::mesh::global is non-null
but the size of the mesh (LocalNx etc.) is not set, because
FieldFactory is created inside BoutMesh::load() via Options.

bendudson committed:
Miscellaneous tidying

Addressing clang-tidy comments and Clang compiler warnings (sprintf use).

bendudson committed:
Field2D/3D: Partly revert 8ebc0a0

Set field sizes in the same way, but get and check sizes on allocation.

bendudson committed:
FieldFactory: Don't test values in Y boundary

bendudson committed:
Clang formatting

bendudson commented:
Probably the easiest way to namespace variables is to append something like "mesh:" (or "input:"?) to the start of the first argument to `addGenerator` here: https://github.com/boutproject/BOUT-dev/blob/field-variable-expression/src/field/field_factory.cxx#L97
i.e.
```
factory.addGenerator(std::string("mesh:" + name, std::make_shared<GridVariable<T>>(&mesh, name));
```
The generator lookup is a simple dictionary key search that happens before any splitting into section names: They only have one namespace.

An alternative is more complex but might simplify code elsewhere: To store fields, or lazily loaded fields, in the `Options` tree and allow them to be used in expressions. Currently only strings can be resolved by `findOption` (https://github.com/boutproject/BOUT-dev/blob/field-variable-expression/src/field/field_factory.cxx#L452) but perhaps that could be extended. Seems like a bigger refactor with more thinking needed.

mikekryjak commented:
I would support accessing the variables by `region::midplane` for example. People may have custom variables in the input file that could clash with the names, and having a prefix makes it easier to identify these variables as well.

ZedThree committed:
Use `gridvar:` namespace for input variables read from grid

ZedThree commented:
I've rebased onto `next` and namespaced grid variables under `gridvar:` -- I figure a new section name like that should prevent accidental collisions with existing variables

bendudson committed:
Merge branch 'next' into field-variable-expression

bendudson committed:
FieldFactory::read_grid_variables Catch null mesh

FieldFactory can be called early in startup, potentially before
the global Mesh has been set. This guards against that by checking
for null mesh passed to `read_grid_variables`.

bendudson committed:
GridVariable::LazyLoad thread safety and performance

LazyLoad::get() always returned a Field3D, so BoutReal or Field2D
grid variables would be converted to Field3D on every call to
generate(). Now converting once to Field3D and storing that.

GridVariable::generate() is called in a BOUT_FOR loop so should
be thread-safe. Adds locks around the `mesh->get` call.

bendudson reviewed (approved):
Thanks @ZedThree !
